litellm/redis_gpt_cache.py

```diff
@@ -0,0 +1,304 @@
+    data_manager = manager_factory(
+        "redis,redis",
+        scalar_params={
+            "redis_host": host,
+            "redis_port": port,
+            "password": password,
+            "global_key_prefix": f"cache_{endpoint_id}_{hashed_llm}",
+        },
+        vector_params={
+            "host": host,
+            "port": port,
+            "password": password,
+            "dimension": embeddings.dimension,
+            "top_k": 1,
+            "collection_name": f"index_{endpoint_id}_{hashed_llm}",
+            "namespace": f"namespace_{endpoint_id}_{hashed_llm}",
+        },
+        eviction_manager="no_op_eviction",
+    )
+
+    eviction_params = {
+        "maxsize": eviction_policy.get("max_size", 100),
+        "policy": eviction_policy.get("policy", "LRU"),
+        "clean_size": int(eviction_policy.get("max_size", 100) * 0.2) or 1,
+        "ttl": eviction_policy.get("ttl"),
+        "on_evict": data_manager._clear,
+    }
+    data_manager.eviction_base = BudServeMemoryCacheEviction(**eviction_params)
+
+    ids = data_manager.s.get_ids(deleted=False)
+    data_manager.eviction_base.put(ids)
+    init_similar_cache(
+        cache_obj=cache_obj,
+        embedding=embeddings,
+        data_manager=data_manager,
+        evaluation=SbertCrossencoderEvaluation(),
+        config=Config(
+            similarity_threshold=cache_config.get(
+                "score_threshold", similarity_threshold
+            ),
+            auto_flush=1,
+        ),
+    )
+
+
+class RedisGPTCache(BaseCache, GPTCache):
+    def __init__(
+        self,
+        host=None,
+        port=None,
+        password=None,
+        redis_url=None,
+        similarity_threshold=None,
+        use_async=False,
+        embedding_model="sentence-transformers/all-MiniLM-L6-v2",
+        **kwargs,
+    ):
+        print_verbose("gptcache redis semantic-cache initializing...")
+        self.similarity_threshold = similarity_threshold
+        self.embedding_model = embedding_model
+        if redis_url is None:
+            # if no url passed, check if host, port and password are passed, if not raise an Exception
+            if host is None or port is None or password is None:
+                # try checking env for host, port and password
+                import os
+
+                host = os.getenv("REDIS_HOST")
+                port = os.getenv("REDIS_PORT")
+                password = os.getenv("REDIS_PASSWORD")
+                if host is None or port is None or password is None:
+                    raise Exception("Redis host, port, and password must be provided")
+
+            redis_url = "redis://:" + password + "@" + host + ":" + port
+        self.redis_params = {"host": host, "port": port, "password": password}
+        GPTCache.__init__(self, init_gptcache_redis)
+        print_verbose(f"gptcache redis semantic-cache redis_url: {redis_url}")
+        self.redis_params = {"host": host, "port": port, "password": password}
```

Comment:
redis_params and GPTCache are initialized twice, please correct

Reply:
noted